[Profiler] Add heuristic to rank events base on computed metrics

davidchencsl opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80098
* __->__ #80094

Test Plan:
Add test in test_profiler.py to check idle time computation.
To test the heuristic, we use serialized real world data stored in test_profiler_utils_golden.py to ensure a run to run stable test.

davidchencsl committed:
Add heuristic to rank events base on computed metrics

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

robieta commented on torch/profiler/_utils.py:
```diff
@@ -1,6 +1,9 @@
 from typing import Dict, List
+import re
+
+import numpy as np
```
While we are compatible with numpy, we don't like to use it as part of the internals because it we want to limit dependencies. (And lots of things fight over numpy version, so we don't want to be part of that brawl during install.) Everything that you're using numpy for you should also be able to just use `torch` functions.

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1288,54 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
+""")
+
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA is required")
+    def test_utils_get_optimizable_events(self):
```
I'm concerned about this test for two reasons:
1) It's vulnerable to flakiness the same way the previous flaky one was
2) It's an E2E test, so it won't catch small changes until something puts it over the edge.

One thing that you might consider doing is running the evaluation just to make sure it runs and maintains the stable invariants (like `len(optimizable_events) == 5`), and then use the results to generate a mock artifact for detailed deterministic testing. If you gate the check on `expecttest.ACCEPT` it will only be used to update, and you can use the `skip` argument to store the literal in a variable.

davidchencsl replied:
I am also not certained on the test for this one:
1. Heuristic changes (cost function change) require to write a  new set of mock test
2. len(optimizable_events) == 5 cannot be guaranteed because we are filtering the events outside downward interval. If due to flakiness we filter out all the event, the invariant is gone too. :(

robieta replied:
1) If we change the cost function won't we just have to update the golden values? (I'm glad you're thinking about that.)
2) Ah, right.

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1288,54 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
+
+        self.assertExpectedInline(
+            res, """\
+0 [CPU (Before cudaLaunchKernel)]
```
Why are these idle times zero? Shouldn't it be the whole duration of the kernel?

robieta reviewed (commented):
Still digesting the main algorithm, but had a couple thoughts on the unit tests.

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py

[ghstack-poisoned]

robieta commented on test/test_profiler_utils_golden.py:
```diff
@@ -0,0 +1,2502 @@
+# Owner(s): ["oncall: profiler"]
```
It seems to me that this would be a lot clearer as just a json and a function to regenerate it in `test_profiler.py`.

robieta commented on test/test_profiler.py:
```diff
@@ -1166,16 +1166,25 @@ def linked_correlation_id(self) -> int:
-            children = []
-            parent = None
+
+            def __init__(self,
```
You don't have to give up the dataclass. The idiomatic way to do non-trivial init on a frozen class is to use `__post_init__` and `object.__set_attr__`.

Example:
https://github.com/pytorch/pytorch/blob/c5f9e043e96f5c60e97292855572576335002c8a/benchmarks/instruction_counts/core/api.py#L62-L66

robieta commented on test/test_profiler.py:
```diff
@@ -1184,28 +1193,71 @@ def end_time_ns(self):
+                MockKinetoEvent("GPU", 1700, 100, 6, 1)
+            ]
+        else:
+            json_list = json.loads(cuda_events_json)
```
I think it would make sense to pull the loading code into a separate class. (That way you could also bundle it with the serialization code.) Right now the flow is a little awkward, because the structure is:
```
[Inline definition]
[Load from file]
[Inline definition]
[Load from file]
```

robieta commented on test/test_profiler.py:
```diff
@@ -1184,28 +1193,71 @@ def end_time_ns(self):
+            cpu_events = []
+            id_map = {}
+            for e in json_list:
+                event = MockProfilerEvent(e['_name'],
```
Is there any reason not to serialize these as a list so you can expand them as positional arguments like you did with `MockKinetoEvent`?

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1341,46 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
+    def test_utils_compute_idle_time(self):
+        profiler = self.generate_mock_profile()
+        basic_evaluation = _utils.BasicEvaluation(profiler)
+        res = ""
```
nit: I have no idea what "res" stands for. Make sure to use descriptive names.

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1341,46 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
+        profiler = self.generate_mock_profile()
+        basic_evaluation = _utils.BasicEvaluation(profiler)
+        res = ""
+        for event_key in basic_evaluation.event_keys:
```
You can make this cleaner with a list comprehension.
```
expected = "\n".join([
    f"{basic_evaluation.metrics[event_key].idle_time_ns} [{event_key.event.name()}]"
    for event_key in basic_evaluation.event_keys])
```

It's also good not to get in the habit of appending strings in a loop, since immutable strings makes your implementation quadratic rather than linear.

davidchencsl replied:
fantastic trick!

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1341,46 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
+            2, print_enable=False)
+        res = ""
+        for event_key in optimizable_events:
+            res += f"{event_key.event.name()}\n"
```
Same comment on adding strings in a loop.

robieta commented on torch/profiler/_utils.py:
```diff
@@ -1,7 +1,9 @@
 from typing import Dict, List
+import re
 
+import torch
```
Don't `import torch` inside library files since it can lead to a circular import.

davidchencsl replied:
But without importing it, how can I use torch.mean, torch.tensor? Do I have to import all of them in?
Checking with other file in torch/profiler, they all did import torch, so that's why I follow the pattern :(

robieta replied:
I take it back. While it isn't ideal, basically every other file also does it so I guess that battle has been lost. If that weren't the case, the way you get around it is by importing outside global scope (e.g. in a function body).

davidchencsl replied:
👍

robieta commented on torch/profiler/_utils.py:
```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
+        event_list = self.rank_events(length)
+        output = ""
+        if len(event_list) == 0:
+            output += "No events to optimize\n"
```
Unused? Generally the way to go about this is something like:
```
output = ["Optimizable events:" if event_list else "No events to optimize"]
for event in event_list:
  output.append(...)
```
One way to think about this is: the more control flow in your function, the easier it is to mess up. If everything follows the same path (and the code works for the degenerate case) it will be more robust.

robieta commented on torch/profiler/_utils.py:
```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
+        qd_values = [e.queue_depth for e in queue_depth_list]
+
+        decrease_interval = []
+        for i in range(len(qd_values) - 1):
```
This is a very deep nested structure and is hard to follow. That's generally a sign that some of the inner blocks should be factored into named functions.

robieta commented on torch/profiler/_utils.py:
```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
+    def get_optimizable_events(self,
+                               length: int = 1,
+                               print_enable: bool = True):
+        event_list = self.rank_events(length)
```
We're no longer using a user provided function?

davidchencsl replied:
I guess since the way that the score is calculated is so involved (using lambda is probably unlikely), the best way for user to define their own function is to inherit from BasicEvaluation, and overwrite the rank_events() ?

robieta replied:
Or have an abstract base class that defines `__call__`. (Which also lets functors easily define their own helper methods and maintain state.)

davidchencsl replied:
I don't quite understand the details here. Is it the case that we let user inherit the ABC, and the __call__ replaces the current rank_events() ? so just a name change?

robieta commented on torch/profiler/_utils.py:
```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
+Source code location: {source_code_location(event.event)}
+Percentage idle time: {self.metrics[event].fraction_idle_time * 100:.2f}%
+{'-'*80}\n"""
+        if print_enable:
```
Why do we unconditionally build `output` if it's only used if print_enable is True? Also, if you find yourself writing a long complex print function that might mean the output should be a class with a `__repr__` function.

davidchencsl replied:
I want to use self.metrics in the print, so there's no easy way for me to put that printing into a class

robieta commented on torch/profiler/_utils.py:
```diff
@@ -46,9 +48,9 @@ def intervals_overlap(self, intervals: List[Interval]):
-            if i + 1 < len(intervals):
-                assert interval.end <= intervals[
-                    i + 1].start, "Intervals must be disjoint"
+            if i + 1 < len(intervals) and interval.end > intervals[i +
```
Oh, sorry. Yes [0, 1]. Same issue though.

davidchencsl replied:
👍

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py to check idle time computation.
To test the heuristic, we use serialized real world data stored in test_profiler_utils_golden.py to ensure a run to run stable test.

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py to check idle time computation.
To test the heuristic, we use serialized real world data stored in test_profiler_utils_golden.py to ensure a run to run stable test.

[ghstack-poisoned]

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py to check idle time computation.
To test the heuristic, we use serialized real world data stored in test_profiler_utils_golden.py to ensure a run to run stable test.

[ghstack-poisoned]

robieta commented on test/test_profiler.py:
```diff
@@ -1215,6 +1235,91 @@ def name(self) -> str:
+    def load_mock_profile():
+        accept = os.getenv("EXPECTTEST_ACCEPT")
+        json_file_path = os.path.dirname(
+            os.path.realpath(__file__)) + "/profiler_utils_mock_events.json"
```
`os.path.join(os.path.realpath(__file__)), "profiler_utils_mock_events.json"))`

robieta commented on torch/profiler/_utils.py:
```diff
@@ -40,15 +41,19 @@ def __eq__(self, other):
+            if i + 1 < len(intervals) and interval.end > intervals[i +
+                                                                   1].start:
+                if interval.end >= intervals[i + 1].end:
+                    del intervals[i + 1]
```
Don't mutate a list while you're iterating over it.

davidchencsl replied:
Change the logic, and added a test to check my sanity

robieta reviewed (approved):
LGTM. (modulo nits) The windows failures don't look related, but probably makes sense to rebase until CI is green.

davidchencsl committed:
Update on "[Profiler] Add heuristic to rank events base on computed metrics"


Test Plan:
Add test in test_profiler.py to check idle time computation.
To test the heuristic, we use serialized real world data stored in test_profiler_utils_golden.py to ensure a run to run stable test.

[ghstack-poisoned]

davidchencsl commented:
@pytorchbot merge
